beacon-chain/sync/data_columns_sampling.go

```diff
@@ -4,24 +4,222 @@ import (
+	for i := uint64(0); i < params.BeaconConfig().NumberOfColumns; i++ {
+		columnToPeerMap[i] = make(map[peer.ID]bool)
+	}
+
+	return &DataColumnSampler1D{
+		p2p:             p2p,
+		peerToColumnMap: make(map[peer.ID]map[uint64]bool),
+	}
+}
+
+// Run implements DataColumnSampler.
+func (d *DataColumnSampler1D) Run(ctx context.Context, stateNotifier statefeed.Notifier) {
+	// initialize peer info first.
+	d.refreshPeerInfo()
+
+	// periodically refresh peer info to keep peer <-> column mapping up to date.
+	async.RunEvery(ctx, PeerRefreshInterval, d.refreshPeerInfo)
+
+	// start the sampling loop.
+	d.samplingLoop(ctx, stateNotifier)
+}
+
+// sampleDataColumns samples data columns from active peers.
+// It should return an error if sampling fails (depends on the actual failing scenario).
+func (d *DataColumnSampler1D) sampleDataColumns(blockRoot [32]byte, samplesCount uint64) error {
+	peerToColumns, err := d.distributeSamplesToPeer(samplesCount)
+	if err != nil {
+		return err
+	}
+
+	eg, ctx := errgroup.WithContext(context.Background())
+	for pid, columns := range peerToColumns {
+		pid, columns := pid, columns
+
+		eg.Go(func() error {
+			return d.sampleDataColumnsFromPeer(ctx, pid, blockRoot, columns)
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
+		log.WithFields(logrus.Fields{
+			"blockRoot":    fmt.Sprintf("%#x", blockRoot),
+			"samplesCount": samplesCount,
+			"error":        err.Error(),
+		}).Error("Failed to sample data columns from peers")
+		return errors.Wrap(err, "error sampling data columns")
+	}
+
+	return nil
+}
+
+// distributeSamplesToPeer dynamically matches samples to peers based on the number of peers the node have.
+// It will try to intelligently choose samples to request from peers that:
+// *. minimizes the chance of false positive result
+// *. maximizes the change to evenly distribute the samples to the peers if possible.
+func (d *DataColumnSampler1D) distributeSamplesToPeer(samplesCount uint64) (map[peer.ID]map[uint64]bool, error) {
+	res := make(map[peer.ID]map[uint64]bool)
+
+	columnsToSample := randomIntegers(samplesCount, params.BeaconConfig().NumberOfColumns)
+	for col := range columnsToSample {
+		peers := d.columnToPeerMap[col]
+		if len(peers) == 0 {
+			return nil, errors.Errorf("no peers responsible for column %d", col)
+		}
+
+		// randomly choose a peer to sample the column.
+		pid := randomPeerFromMap(peers)
+		if _, ok := res[pid]; !ok {
+			res[pid] = make(map[uint64]bool)
+		}
+		res[pid][col] = true
+	}
+
+	return res, nil
+}
+
+// sampleDataColumnsFromPeer samples data columns from a peer.
+func (d *DataColumnSampler1D) sampleDataColumnsFromPeer(ctx context.Context, pid peer.ID, blockRoot [32]byte, columns map[uint64]bool) error {
+	// SendDataColumnSidecarByRoot()
+
+	// dataColumnIdentifiers := make(types.BlobSidecarsByRootReq, 0, len(columns))
+	panic("not implemented")
+}
+
+// Refresh peer information.
+func (d *DataColumnSampler1D) refreshPeerInfo() {
```

Comment:
This function creates a memory leak: Peers are only added, and never removed.
You need to add a prune mechanism, where peers not seen since a certain amount of time are removed from the maps.